File: jscomp/stdlib/complex.ml
(***********************************************************************)
(*                                                                     *)
(*                                OCaml                                *)
(*                                                                     *)
(*            Xavier Leroy, projet Cristal, INRIA Rocquencourt         *)
(*                                                                     *)
(*  Copyright 2002 Institut National de Recherche en Informatique et   *)
(*  en Automatique.  All rights reserved.  This file is distributed    *)
(*  under the terms of the GNU Library General Public License, with    *)
(*  the special exception on linking described in file ../LICENSE.     *)
(*                                                                     *)
(***********************************************************************)

(* Complex numbers *)

type t = { re: float; im: float }

let zero = { re = 0.0; im = 0.0 }
let one = { re = 1.0; im = 0.0 }
let i = { re = 0.0; im = 1.0 }

let add x y = { re = x.re +. y.re; im = x.im +. y.im }

let sub x y = { re = x.re -. y.re; im = x.im -. y.im }

let neg x = { re = -. x.re; im = -. x.im }

let conj x = { re = x.re; im = -. x.im }

let mul x y = { re = x.re *. y.re -. x.im *. y.im;
                im = x.re *. y.im +. x.im *. y.re }

let div x y =
  if abs_float y.re >= abs_float y.im then
    let r = y.im /. y.re in
    let d = y.re +. r *. y.im in
    { re = (x.re +. r *. x.im) /. d;
      im = (x.im -. r *. x.re) /. d }
  else
    let r = y.re /. y.im in
    let d = y.im +. r *. y.re in
    { re = (r *. x.re +. x.im) /. d;
      im = (r *. x.im -. x.re) /. d }

let inv x = div one x

let norm2 x = x.re *. x.re +. x.im *. x.im

let norm x =
  (* Watch out for overflow in computing re^2 + im^2 *)
  let r = abs_float x.re and i = abs_float x.im in
  if r = 0.0 then i
  else if i = 0.0 then r
  else if r >= i then
    let q = i /. r in r *. sqrt(1.0 +. q *. q)
  else
    let q = r /. i in i *. sqrt(1.0 +. q *. q)

let arg x = atan2 x.im x.re

let polar n a = { re = cos a *. n; im = sin a *. n }

let sqrt x =
  if x.re = 0.0 && x.im = 0.0 then { re = 0.0; im = 0.0 }
  else begin
    let r = abs_float x.re and i = abs_float x.im in
    let w =
      if r >= i then begin
        let q = i /. r in
        sqrt(r) *. sqrt(0.5 *. (1.0 +. sqrt(1.0 +. q *. q)))
      end else begin
        let q = r /. i in
        sqrt(i) *. sqrt(0.5 *. (q +. sqrt(1.0 +. q *. q)))
      end in
    if x.re >= 0.0
    then { re = w;  im = 0.5 *. x.im /. w }
    else { re = 0.5 *. i /. w;  im = if x.im >= 0.0 then w else -. w }
  end

let exp x =
  let e = exp x.re in { re = e *. cos x.im; im = e *. sin x.im }

let log x = { re = log (norm x); im = atan2 x.im x.re }

let pow x y = exp (mul y (log x))

File: lib/es6/complex.js



var one = {
  re: 1.0,
  im: 0.0
};

function add(x, y) {
  return {
          re: x.re + y.re,
          im: x.im + y.im
        };
}

function sub(x, y) {
  return {
          re: x.re - y.re,
          im: x.im - y.im
        };
}

function neg(x) {
  return {
          re: -x.re,
          im: -x.im
        };
}

function conj(x) {
  return {
          re: x.re,
          im: -x.im
        };
}

function mul(x, y) {
  return {
          re: x.re * y.re - x.im * y.im,
          im: x.re * y.im + x.im * y.re
        };
}

function div(x, y) {
  if (Math.abs(y.re) >= Math.abs(y.im)) {
    var r = y.im / y.re;
    var d = y.re + r * y.im;
    return {
            re: (x.re + r * x.im) / d,
            im: (x.im - r * x.re) / d
          };
  }
  var r$1 = y.re / y.im;
  var d$1 = y.im + r$1 * y.re;
  return {
          re: (r$1 * x.re + x.im) / d$1,
          im: (r$1 * x.im - x.re) / d$1
        };
}

function inv(x) {
  return div(one, x);
}

function norm2(x) {
  return x.re * x.re + x.im * x.im;
}

function norm(x) {
  var r = Math.abs(x.re);
  var i = Math.abs(x.im);
  if (r === 0.0) {
    return i;
  }
  if (i === 0.0) {
    return r;
  }
  if (r >= i) {
    var q = i / r;
    return r * Math.sqrt(1.0 + q * q);
  }
  var q$1 = r / i;
  return i * Math.sqrt(1.0 + q$1 * q$1);
}

function arg(x) {
  return Math.atan2(x.im, x.re);
}

function polar(n, a) {
  return {
          re: Math.cos(a) * n,
          im: Math.sin(a) * n
        };
}

function sqrt(x) {
  if (x.re === 0.0 && x.im === 0.0) {
    return {
            re: 0.0,
            im: 0.0
          };
  }
  var r = Math.abs(x.re);
  var i = Math.abs(x.im);
  var w;
  if (r >= i) {
    var q = i / r;
    w = Math.sqrt(r) * Math.sqrt(0.5 * (1.0 + Math.sqrt(1.0 + q * q)));
  } else {
    var q$1 = r / i;
    w = Math.sqrt(i) * Math.sqrt(0.5 * (q$1 + Math.sqrt(1.0 + q$1 * q$1)));
  }
  if (x.re >= 0.0) {
    return {
            re: w,
            im: 0.5 * x.im / w
          };
  } else {
    return {
            re: 0.5 * i / w,
            im: x.im >= 0.0 ? w : -w
          };
  }
}

function exp(x) {
  var e = Math.exp(x.re);
  return {
          re: e * Math.cos(x.im),
          im: e * Math.sin(x.im)
        };
}

function log(x) {
  return {
          re: Math.log(norm(x)),
          im: Math.atan2(x.im, x.re)
        };
}

function pow(x, y) {
  return exp(mul(y, log(x)));
}

var zero = {
  re: 0.0,
  im: 0.0
};

var i = {
  re: 0.0,
  im: 1.0
};

export {
  zero ,
  one ,
  i ,
  neg ,
  conj ,
  add ,
  sub ,
  mul ,
  inv ,
  div ,
  sqrt ,
  norm2 ,
  norm ,
  arg ,
  polar ,
  exp ,
  log ,
  pow ,
  
}
/* No side effect */
